docs/src/solvers.md

```diff
@@ -1,7 +1,84 @@
 Interacting with solvers
 ========================
 
-TODO: Describe the connection between JuMP and solvers. Automatic vs. Manual
-mode. CachingOptimizer. How to set/change solvers. How to set parameters (solver
+A JuMP model stores a
+[MathOptInterface (MOI)](https://github.com/JuliaOpt/MathOptInterface.jl)
+backend internally that contains the optimization solver. The JuMP layer on top
+of this MOI backend is aimed to be as lightweight as possible:
+
+* JuMP does not maintain any copy of the model outside this MOI backend.
+* JuMP variable (resp. constraint) references are simple structures containing
+  both a reference to the JuMP model and the MOI index of the variable (resp.
+  constraint).
+* JuMP gives the constraints to the MOI backend in the form provided by the user
+  without doing any automatic reformulation.
+* variables additions, constraints additions/modifications and objective
+  modifications are directly applied to the MOI backend thus expecting the
+  backend to support such modifications.
+
+While this allows JuMP API to to be a thin wrapper on top of the solver API,
+as mentionned in the last point above, this seems rather demanding on the
+solver. Indeed, while some solvers support incremental building of the model and
+modifications before and after solve, other solvers only support the model being
+copied at once before solve. Moreover it seems to require all solvers to
+implement all possible reformulations independently which seems both very
+ambitious and might generate a lot of duplicated code.
+
+These apparent limitations are in fact addressed at the MOI level in a manner
+that is completely transparent to JuMP. While the MOI API may seem very
+demanding, it allows MOI models to be a succession of lightweight MOI layers
+that fill the gap between JuMP requirements and the solver capabilities.
+
+JuMP models can be created in three different modes: Automatic, Manual and
+Direct.
+
+## Automatic and Manual modes
+
+In Automatic and Manual modes, two MOI layers are automatically applied to the
+optimizer:
+
+* `CachingOptimizer`: it maintain a cache of the model so that when the
+  the optimizer does not support a modification of the model, the optimizer's
```

Comment:
"incremental change" seems more appropriate than "modification" (since it could include adding constraints)